src/osmet.rs

```diff
@@ -0,0 +1,969 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::ffi::OsStr;
+use std::fs::{File, OpenOptions};
+use std::io::{self, BufReader, BufWriter, ErrorKind, Read, Seek, SeekFrom, Write};
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::os::unix::fs::FileTypeExt;
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+
+use error_chain::bail;
+use nix::mount;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use tempfile::NamedTempFile;
+use walkdir::WalkDir;
+use xz2::read::XzDecoder;
+use xz2::write::XzEncoder;
+
+use crate::blockdev::*;
+use crate::cmdline::*;
+use crate::errors::*;
+use crate::fiemap::*;
+
+pub fn osmet_fiemap(config: &OsmetFiemapConfig) -> Result<()> {
+    println!("{:?}", fiemap_path(config.file.as_str().as_ref())?);
+    Ok(())
+}
+
+// The sha2 crate wants us to use generic-array and typenum to deal with checksums, which are ways
+// to hack around Rust's lack of const generics. We sidestep that goop here in favour of the simple
+// [u8; 32] type since it's cleaner and we only have to deal with SHA256s.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct Sha256Digest([u8; 32]);
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Mapping {
+    extent: Extent,
+    object: Sha256Digest,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetPartition {
+    start_offset: u64,
+    end_offset: u64,
+    mappings: Vec<Mapping>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Osmet {
+    /// Partitions for which we've registered mappings.
+    partitions: Vec<OsmetPartition>,
+    /// Checksum of the final disk image. Used for unpacking verification.
+    checksum: Sha256Digest,
+    /// Size of the final disk image. Used for unpacking verification.
+    size: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetFile {
+    magic: [u8; 8],
+    major: u16,
+    minor: u16,
+    osmet: Osmet,
+}
+
+const OSMET_FILE_HEADER_MAGIC: [u8; 8] = *b"OSMET\0\0\0";
+
+pub fn osmet_pack(config: &OsmetPackConfig) -> Result<()> {
+    // First, mount the two main partitions we want to suck out data from: / and /boot. Note
+    // MS_RDONLY; this also ensures that the partition isn't already mounted rw elsewhere.
```

Comment:
Best practice here would be for the invoking program to use `unshare -m` right?

Reply:
Yes, agreed. Hmm, and I guess since this has to run as root, we could fold the unshare logic here. Though will leave that as a follow up.

Reply:
Split that out into https://github.com/coreos/coreos-installer/issues/208.